src/components/UsernameInput/UsernameInput.tsx

```diff
@@ -13,6 +12,19 @@ interface UsernameInputProps extends Omit<TextStyleProps, "onColorChange"> {
 }
 
 const UsernameInput: FC<UsernameInputProps> = ({ state, onInputChange, label }) => {
+    const [emailValidation, setEmailValidation] = useState<"valid" | "invalid">();
+    const emailRegex = new RegExp(/^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}$/);
+
+    useEffect(() => {
+        if (state !== undefined && state.variant === "email") {
+            validateEmail(state.textValue);
+        }
+    }, [state.textValue]);
```

Comment:
While this works, we could get around it without depending on a `useEffect`

we could move this to inside of the `handleInputChange` method 

```javascript
    const handleInputChange = (value: string) => {
        onInputChange({ variant: state.variant, textValue: value });
        if(state.variant === 'email') { 
           validateEmail(value);
        }
    };
```

this way we _**guarantee**_ that the validation check will occur every time the input changes instead of depending on a useEffect to trigger it as a side effect

Reply:
🤔 ok ok implementing without the useEffect hook!